### PR TITLE
Ensure that symbols exposed to a BinderFactory are going to survive a race between threads.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:945]: Do we want to keep this assert for `nonTypeMembers` instead?

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:945]: ie, if `HaveIndexers` is true, there should be an indexer.

---
In reply to: [568845087](https://github.com/dotnet/roslyn/pull/50961#discussion_r568845087) [](ancestors = 568845087)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1395]: >s => s.Name [](start = 87, length = 11)

Nit: consider adding `static` to these lambdas (and below).

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1450]: >(declared.NonTypeMembers.Contains(m => m == (object)member) || declared.RecordPrimaryConstructor == (object)member)) || [](start = 39, length = 119)

Nit: considering putting this on the next line to make it clear where the parenthesized expression starts and ends.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1450]: Actually, I misread. This is fine.

---
In reply to: [568850528](https://github.com/dotnet/roslyn/pull/50961#discussion_r568850528) [](ancestors = 568850528)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1395]: > Nit: consider adding static to these lambdas (and below).

I see no benefit in doing so for simple lambdas like that.

---
In reply to: [568848641](https://github.com/dotnet/roslyn/pull/50961#discussion_r568848641) [](ancestors = 568848641)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2875]: Consider adding a remark to this method that `membersByName` should only be a ref to a local, or corruption could result.

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:2875]: > Consider adding a remark to this method that membersByName should only be a ref to a local, or corruption could result.

The statement in the suggestion is false.

---
In reply to: [568865128](https://github.com/dotnet/roslyn/pull/50961#discussion_r568865128) [](ancestors = 568865128)

**Review by 333fred — APPROVED** [review]
LGTM (commit 1). Just a couple of small comments.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1448]: The lack of a `Volatile.Read` caught my eye reading this method. There are a number of other places in the code where we access this member via `Volatile.Read`. I presume that's because we are accounting for the fact the current thread isn't guaranteed to hit one of the paths where we `CompareExchange` the value and hence could get a stale read here. 

The one call to this method is from the `Binder` and I'm unsure if that is on a path where we call through `BuildMembersAndInitializers`. If not I think we need a `Volatile.Read` here. Or am I missing another way that it's accounted for on this code path?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1448]: There is only one call site for this function and it is not in ```Binder```. Regardless, it wouldn't hurt to add some volatile reads here to avoid any possibility of a false assert. I'll make this adjustment in the next PR. 

---
In reply to: [569945538](https://github.com/dotnet/roslyn/pull/50961#discussion_r569945538) [](ancestors = 569945538)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler Please review, need a second sign-off.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1445]: Consider adding a comment here elaborating on what `Exposure` means (basically what invariant are we trying to maintain).

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (jaredpar) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1451]: Think we should use a `Volatile.Read` here for similar reasons.

**Review by jaredpar — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@jaredpar, @333fred, @dotnet/roslyn-compiler Since 'master' is in lockdown now, I decided to add another commit to this PR instead of waiting when we can merge this PR and then submitting another PR. I already have two sign-offs for the first commit, please review the second commit, or both.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1457]: >? [](start = 64, length = 1)

Since this assert expects `_lazyTypeMembers` to be non-null, you could just `!` away the nullable warning and simplify the condition.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1460]: >|| member is [](start = 51, length = 12)

Nit: consider just using `or`

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (333fred) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1468]: Not sure I understand this one. Can you add a comment to explain why we need to detour to the event?

**Comment by 333fred** [comment]
Done review pass (commit 2). Couple of small questions before I sign off.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1457]: > Since this assert expects _lazyTypeMembers to be non-null, you could just ! away the nullable warning and simplify the condition.

I do not find this condition particularly complex and I didn't write it this way because a nullable warning. It naturally happened written this way. Also, it is not my goal here to assume something, and crash with null ref when the assumption is violated.  

---
In reply to: [571282589](https://github.com/dotnet/roslyn/pull/50961#discussion_r571282589) [](ancestors = 571282589)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1468]: > Not sure I understand this one. Can you add a comment to explain why we need to detour to the event?

Fields like that are not added to members list. I can add a comment, but prefer doing that in a separate PR, unless there are other reasons to make changes in this one. 

---
In reply to: [571285867](https://github.com/dotnet/roslyn/pull/50961#discussion_r571285867) [](ancestors = 571285867)

**Review by 333fred — APPROVED** [review]
LGTM (commit 2)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:1468]: Perfectly fine as a followup.

---
In reply to: [571299314](https://github.com/dotnet/roslyn/pull/50961#discussion_r571299314) [](ancestors = 571299314,571285867)

**Review by jaredpar — APPROVED** [review]

**Pull request merged**
